### PR TITLE
Support repeated measurement keys in cirq.Result

### PR DESCRIPTION
**Pull request by maffoo**
Part of #4274

### CODE REVIEW

**Comment by maffoo** [comment]
I've rebased this on master including the separation of `cirq.Result` into an ABC. One design question here is what the new interface for accessing 3D arrays of repeated measurements should look like. I've called in `measurements2` as a placeholder. Any thoughts @95-martin-orion?

**Comment by 95-martin-orion** [comment]
> One design question here is what the new interface for accessing 3D arrays of repeated measurements should look like.

Let's cover this at our meeting Thursday - there are a bunch of open threads here and I'd like to get them woven together before rushing ahead on this.

File: cirq-core/cirq/work/zeros_sampler.py
Comment (95-martin-orion) [inline cirq-core/cirq/work/zeros_sampler.py:71]: Two things:
- This also needs to check that the dimensions for each qudit are the same - e.g. (3, 2) != (2, 3).
- This feels like the wrong place for this check to live, since presumably every sampler/simulator is going to require it. Could we move it up to the Sampler superclass?

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
A pretty self-explanatory implementation of `Result` as described in the RFC, though there's a comment from @Strilanc there you should take a look at before we commit to this.

**Comment by 95-martin-orion** [comment]
@maffoo Heads up: due #4899 requiring access to `Result.records` and needing to be submitted before the v0.14 in a couple weeks, we need to merge this PR as soon as possible.

File: cirq-core/cirq/work/zeros_sampler.py
Comment (maffoo) [reply cirq-core/cirq/work/zeros_sampler.py]: Moved this into a helper function `Sampler._get_measurement_shapes` in the base class. Note that this is basically a preliminary version of the "measurement_key_shape" protocol proposed in https://github.com/quantumlib/Cirq/issues/4888.

**Review by 95-martin-orion — APPROVED** [review]

**Comment by 95-martin-orion** [comment]
Approved, but requires conflict resolution before merging. The change looks fairly minor.

**Comment by daxfohl** [comment]
I noticed one surprising (to me anyway) behavior of `ResultDict` is that, when passing in `records` but not `measurements`, and *any* measurement occurs more than once, then `ResultDict.measurements['key']` fails even if `key` was only measured once.

**Pull request merged**
